### PR TITLE
Add PR #75 review feedback for python files from @MSeal and @willingc

### LINKED ISSUE #86: Address feedback from PR #75
Address open feedback from #75

### PR DESCRIPTION
**Pull request by willingc**
- Addresses #75 feedback for Python files
- Partially closes #86

### CODE REVIEW

File: bookstore/__init__.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
+# TODO: refactor the import, PACKAGE_DIR, del
```
Comment (MSeal) [inline bookstore/__init__.py]: Let's just remove the PACKAGE_DIR from this file. We can just move it to where it's used in `clone.py`

File: bookstore/client/nb_client.py [hunk 2]
```diff
@@ -30,6 +30,7 @@
 
 
 def extract_kernel_id(connection_file):
+    # TODO should this revert to not using regex ?
```
Comment (MSeal) [inline bookstore/client/nb_client.py]: Yes.

File: bookstore/client/store_client.py [hunk 3]
```diff
@@ -7,37 +7,38 @@
...
 
-    def __init__(self, s3_bucket=""):
+    def __init__(self, s3_bucket=None):
         if s3_bucket:
             self.default_bucket = s3_bucket
         super().__init__()
 
     @property
     def publish_endpoint(self):
+        # TODO: refactor to /api/bookstore/publish/ ?
```
Comment (MSeal) [inline bookstore/client/store_client.py]: Agreed. Can be a separate PR.

File: bookstore/__init__.py (hunk 1 above)
Comment (mpacer) [reply bookstore/__init__.py]: I generally prefer to do this at the `__init__.py` so that it can be accessible from other locations, including those outside of the library.

Could you explain the reason you disprefer this pattern?

File: bookstore/client/nb_client.py (hunk 2 above)
Comment (mpacer) [reply bookstore/client/nb_client.py]: No I disagree. The previous version was riddled with bugs. Regex is the only clean way to extract a prefix in python. `connection_filename.lstrip("kernel-") `  would strip all instances of the characters "e" "k" "r" "n" "l" and "-". So, when the connection file began with an "e" this would result in being unable to find the connection file.

File: bookstore/client/nb_client.py [hunk 4]
```diff
@@ -81,21 +82,21 @@ class NotebookSession:  # (NamedTuple):
...
     #     kernel: KernelInfo
     #     notebook: dict # deprecated API {'path': 'Untitled38.ipynb', 'name': ''}}}
 
-    def __init__(self, *args, path, name, type, kernel, notebook, **kwargs):
+    def __init__(self, *args, path, name, kind, kernel, notebook, **kwargs):
         self.model = {
             "path": path,
             "name": name,
-            "type": type,
+            "kind": kind,
```
Comment (mpacer) [inline bookstore/client/nb_client.py]: This no longer is able to accept the literal response from the NotebookSession API… to not have this defeats the purpose of the type class. 
This has the form:
https://github.com/jupyter/notebook/blob/01b084381c367bdca1d58b176428b728a67628ca/notebook/services/api/api.yaml#L761-L778
```
  Session:
    description: A session
    type: object
    properties:
      id:
        type: string
        format: uuid
      path:
        type: string
        description: path to the session
      name:
        type: string
        description: name of the session
      type:
        type: string
        description: session type
      kernel:
        $ref: '#/definitions/Kernel'
```

File: bookstore/client/store_client.py [hunk 5]
```diff
@@ -7,37 +7,38 @@
...
         api_endpoint = "/api/bookstore/published/"
         return f"{self.url}{api_endpoint}"
 
     def publish(self, path=None):
         """Publish notebook to bookstore"""
         if path is None:
             path = self.session.path
-        nb_json = self.get_contents(self.session.path)['content']
+        nb_json = self.get_contents(path)['content']
```
Comment (mpacer) [inline bookstore/client/store_client.py]: thank you for catching that bug!

File: bookstore/client/store_client.py (hunk 3 above)
Comment (mpacer) [reply bookstore/client/store_client.py]: This is specifically a request from @rgbkrk to have this be more REST-ful in its naming. Rather than "publish" the action this is intended to be a reference to a location "published" that refers to the notebooks that are there.

I find it somewhat confusing as well, but I don't want to break from this approach without getting @rgbkrk's sign off.

File: bookstore/client/store_client.py [hunk 6]
```diff
@@ -7,37 +7,38 @@
...
+        target_url = f"{self.publish_endpoint}{path}"
 
-        resp = self.req_session.put(target_url, json=json_body)
-        return resp
+        response = self.req_session.put(target_url, json=json_body)
+        return response
 
     @property
     def clone_endpoint(self):
+        # TODO: refactor to /api/bookstore/clone/ ?
```
Comment (mpacer) [inline bookstore/client/store_client.py]: Same comments as above around "published"… this is specifically a request from @rgbkrk to have this be more REST-ful in its naming. Rather than "clone" the action, this is intended to be a reference to a location "cloned" that refers to the notebooks that are there.

I find this even more confusing than the `/published` name, but, again, I don't want to break from this approach without getting @rgbkrk's sign off.

File: bookstore/clone.py [hunk 7]
```diff
@@ -24,24 +22,24 @@ def initialize(self):
...
         self.session = aiobotocore.get_session()
 
-    async def _clone(self, s3_bucket, file_key):
-        path = file_key
+    async def _clone(self, s3_bucket, s3_key_file):
 
         self.log.info(f"bucket: {s3_bucket}")
-        self.log.info(f"key: {file_key}")
-        full_s3_path = s3_path(s3_bucket, path)
+        self.log.info(f"key: {s3_key_file}")
```
Comment (mpacer) [inline bookstore/clone.py]: I think this should probably be `s3_file_key` or `s3_object_key` rather than `s3_key_file`.

File: bookstore/publish.py [hunk 8]
```diff
@@ -1,16 +1,12 @@
...
-from notebook.base.handlers import path_regex
+from notebook.base.handlers import APIHandler, path_regex
 from notebook.utils import url_path_join
 from tornado import web
 
-from .s3_paths import s3_path
-from .s3_paths import s3_key
-from .s3_paths import s3_display_path
 from .bookstore_config import BookstoreSettings
+from .s3_paths import s3_display_path, s3_key, s3_path
```
Comment (mpacer) [inline bookstore/publish.py]: I had split these out because of your previous styling edit that took my one line imports and split them into multiple. I prefer them to be on one line, but was trying to incorporate your changes.

Can you give me guidance on when to go in one direction or the other?

File: ci/clone_request.py [hunk 9]
```diff
@@ -33,24 +35,11 @@ def post(**kwargs):
...
-
-    # tests minio created 'bookstore' bucket
-    # res = get("?s3_bucket=bookstore&s3_key=ci-published/ci-published.ipynb")
-    # print(res.content)
-    # res = post(s3_bucket="nteract-notebooks", s3_key="Introduction_to_Chainer.ipynb")
-    # print(res.content)
-    res = post(s3_bucket="bookstore", s3_key="ci-published/ci-published.ipynb")
-    pprint.pprint(res.json())
-    res = post(
+    response = post(s3_bucket="bookstore", s3_key="ci-published/ci-published.ipynb")
```
Comment (mpacer) [inline ci/clone_request.py]: The comments are purposefully there to illustrate other things that are intended to be easily tested interactively. They should not be removed.

Comment (mpacer) [reply ci/clone_request.py]: Or rather…if we are going to remove these comments, we might as well remove the file in its entirety as it's role is in aiding debugging. I thought including it in the repo would be useful for others to have as an example. 

To be honest, the motivation of the client was to make it easier to do this kind of quick debugging.

**Comment by mpacer** [comment]
@MSeal re: your comments on testing. 

I would love your insight on how to test the various clients. I am super enthusiastic about that just had a tricky time getting that to start as to do it properly requires some pretty extensive integration test work. 

Regarding the cloning functionality itself that was introduced in #75, that was what the js based integration tests were testing.

File: ci/clone_request.py (hunk 9 above)
Comment (willingc) [reply ci/clone_request.py]: Honestly, I thought this was uncommented code not a comment for the user of the library. It would be better to perhaps put in a docstring instead and make it clear that these are tips for debugging.

File: bookstore/client/nb_client.py (hunk 4 above)
Comment (MSeal) [reply bookstore/client/nb_client.py]: Then maybe we should leave "type" as a key but make the api function signature rename the argument variable. Having a reserved word variable name is problematic in python.

File: bookstore/publish.py (hunk 8 above)
Comment (MSeal) [reply bookstore/publish.py]: I tend to let the linter do whatever the linter likes unless it's really jarring. I generally leave all import from the same path in the same statement and let it spill to multi-line over separate repeated imports. Otherwise it's easy to miss duplicate imports and adds confusion around what importing is doing.Though that's kjust my opinion on the topic.

File: bookstore/clone.py (hunk 7 above)
Comment (MSeal) [reply bookstore/clone.py]: It's odd that we force keys to be passed around rather than keeping credential collections purely within the s3 client. This makes setups where there's no key file unable to use this tool despite having valid access to S3.

**Comment by MSeal** [comment]
@mpacer I would argue we need unittests first here most of all. You'll need to mock s3 and web interfaces above and below the various functions, but there's a lot of logic here for handling edge cases that need coverage for reliable development. A starting place would be to get a test for clone `get` receiving and empty file path. I can help set up first first couple tests when I get back if it'd help get the ball rolling.

File: bookstore/client/nb_client.py (hunk 4 above)
Comment (mpacer) [reply bookstore/client/nb_client.py]: The function signature is what needs to have it. I'd be fine making it an attribute that has a different name (it's less ideal & goes against the goals of the API… but possible), but being able to unpack the json response without altering it is the specific intent of this API. 

Also, it's technically not a reserved keyword… just a super commonly used function that is not best practice to overwrite. Keywords: https://docs.python.org/3/reference/lexical_analysis.html#keywords. Otherwise this would be a syntax error.

I don't like doing this either and if it weren't for the explicit desire to be isomorphic to the response object I wouldn't push on this.

File: bookstore/clone.py (hunk 7 above)
Comment (mpacer) [reply bookstore/clone.py]: these aren't s3_keys. That's what I'm saying. They are paths to the object we should come up with a better name for them as they clearly just lead to confusion.

File: bookstore/publish.py (hunk 8 above)
Comment (mpacer) [reply bookstore/publish.py]: I agree, I prefer to let the linter decide, but I was trying to incorporate the changes @willingc  had made in this PR: https://github.com/nteract/bookstore/pull/70/files#diff-b79c22e74e3095962bbd282b0bd0c040R13

File: ci/clone_request.py (hunk 9 above)
Comment (mpacer) [reply ci/clone_request.py]: If their purpose is that ambiguous shouldn't we just remove them from the library?

File: bookstore/clone.py (hunk 7 above)
Comment (mpacer) [reply bookstore/clone.py]: Ah perhaps we should call it s3_object_key? that's the language they use in the s3 docs: https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-keys

Comment (willingc) [reply bookstore/clone.py]: I could get behind `s3_object_key_name` which seems to accurately reflect the AWS docs and not imply that we are passing around keys or files.

File: bookstore/client/store_client.py (hunk 6 above)
Comment (MSeal) [reply bookstore/client/store_client.py]: Let's change it, there was general consensus from folks that the action routes shouldn't be past-tensed, but retrieval routes can be nouns. If it's generating the object let's leave it a present tense, it it's grabbing older generated objects then the past tense or pluralized nouns is probably correct.

File: bookstore/client/nb_client.py (hunk 4 above)
Comment (MSeal) [reply bookstore/client/nb_client.py]: Just wanted to raise it as it can cause unexpected issues in the future. Not a blocker or required change.

File: bookstore/clone.py (hunk 7 above)
Comment (mpacer) [reply bookstore/clone.py]: I went with `s3_object_key` as that is the way that it is referred to most generally in the docs and is a slight bit shorter. Should we change the url param to match? Since it hasn't been officially released yet, now is probably the time to do it.

File: bookstore/client/store_client.py (hunk 6 above)
Comment (mpacer) [reply bookstore/client/store_client.py]: Ok, a change PR is underway.

File: bookstore/clone.py (hunk 7 above)
Comment (willingc) [reply bookstore/clone.py]: Yep. I would make it consistent. Thanks @mpacer

File: bookstore/client/store_client.py (hunk 3 above)
Comment (rgbkrk) [reply bookstore/client/store_client.py]: It's probably ok to just refer to it in an RPC way, especially since I'm the only one who thought otherwise. I'd prefer what makes sense to most people.

Comment (willingc) [reply bookstore/client/store_client.py]: I think at this point that we should leave as is.

File: bookstore/__init__.py (hunk 1 above)
Comment (willingc) [reply bookstore/__init__.py]: I have removed the TODO to move this forward. We can always refactor at a later time if needed.

File: ci/clone_request.py (hunk 9 above)
Comment (willingc) [reply ci/clone_request.py]: I've updated the docstrings so that the content remains and is more understandable.

File: bookstore/clone.py (hunk 7 above)
Comment (willingc) [reply bookstore/clone.py]: I changed it in this PR.

**Comment by willingc** [comment]
@mpacer I spent some time this morning going through this PR in detail. All tests are passing (thanks to M's fix to the js ci - changing const to let). Let's merge this and move forward.

We're getting very close to a release. Just a few more tests and a changelog after this lands.

File: ci/utils.js [hunk 10]
```diff
@@ -9,7 +9,7 @@ const url_path_join = function(...pieces) {
   // """
   const initial = pieces[0].startsWith("/");
   const final = pieces[pieces.length - 1].endsWith("/");
-  const result = pieces
+  let result = pieces
```
Comment (mpacer) [inline ci/utils.js]: I am a little confused — shouldn't this already be changed on master?

https://github.com/nteract/bookstore/blob/ad44c3abad19dce0a834e320cd30a5b18f2ccc29/ci/utils.js#L10-L15

Comment (willingc) [reply ci/utils.js]: It is but this predated that so it is probably a merge conflict now that needs to be taken care of in this PR so that master is used.

File: bookstore/client/nb_client.py [hunk 11]
```diff
@@ -81,21 +82,21 @@ class NotebookSession:  # (NamedTuple):
     #     id: str #'68d9c58f-c57d-4133-8b41-5ec2731b268d',
     #     path: str #'Untitled38.ipynb',
     #     name: str #'',
-    #     type: str #'notebook',
+    #     kind: str #'notebook',
     #     kernel: KernelInfo
     #     notebook: dict # deprecated API {'path': 'Untitled38.ipynb', 'name': ''}}}
 
-    def __init__(self, *args, path, name, type, kernel, notebook, **kwargs):
+    def __init__(self, *args, path, name, kind, kernel, notebook, **kwargs):
```
Comment (mpacer) [inline bookstore/client/nb_client.py]: I would prefer that we don't change this as it would make it so that we cannot directly consume the contents of the API response from the `notebook`'s `/api/sessions/` in https://github.com/nteract/bookstore/pull/90/files#diff-7d77120266750554a7e864156f790108R139

File: bookstore/client/nb_client.py [hunk 12]
```diff
@@ -30,6 +30,7 @@
 
 
 def extract_kernel_id(connection_file):
+    # regex is used as a more robust approach than lstrip
```
Comment (mpacer) [inline bookstore/client/nb_client.py]: Many thanks for the clarifying comment! Looks great :D!

File: ci/utils.js (hunk 10 above)
Comment (willingc) [reply ci/utils.js]: Rebased

**Comment by willingc** [comment]
@mpacer I think that this has all the changes that we discussed on our call.
